### PR TITLE
In-source zinc's LM integration code

### PR DESCRIPTION
**Pull request by dwijnand**
This requires a new version of Zinc in order to build (tested using local source dependencies).

### CODE REVIEW

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1537/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/4617/head-pull/658/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1537/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@9674e4a6e9c45eab96331a9af21715255d26179c
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/658">pull/658/head       </A> | sbt/zinc@0b6470f2b8558e5c8b77e2942f02aca19e3bbad8
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@ca966bcd6c73400b11c7a78e8604ea211f32e3d8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@473d42ac782f9449dbd7c7940a67208abbff1861

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=pull/658/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1538/console">is in progress...</a>

**Comment by eed3si9n** [comment]
Could we keep git history somehow?

**Comment by dwijnand** [comment]
Sure, I'll do the same thing I did in https://github.com/sbt/sbt/pull/3302.

I'll probably get round to it in a few days.

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/4617/head-pull/658/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1538/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@7466ee4913d5281b99130c217c7bac02347adb76
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/658">pull/658/head       </A> | sbt/zinc@0b6470f2b8558e5c8b77e2942f02aca19e3bbad8
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@ca966bcd6c73400b11c7a78e8604ea211f32e3d8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@473d42ac782f9449dbd7c7940a67208abbff1861

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=pull/658/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1539/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/4617/head-pull/658/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1539/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@7466ee4913d5281b99130c217c7bac02347adb76
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/658">pull/658/head       </A> | sbt/zinc@0b6470f2b8558e5c8b77e2942f02aca19e3bbad8
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@ca966bcd6c73400b11c7a78e8604ea211f32e3d8
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@473d42ac782f9449dbd7c7940a67208abbff1861

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=pull/658/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by dwijnand** [comment]
```
[sbt212] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt212] [warn] 	::          UNRESOLVED DEPENDENCIES         ::
[sbt212] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt212] [warn] 	:: org.scala-sbt#compiler-bridge_2.13.0-M5;1.3.0-bin-20190418T141950: not found
[sbt212] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt212] [info] ZincComponentCompilerSpec:
[sbt212] [info] - should compile the bridge for Scala 2.10.5 and 2.10.6 *** FAILED ***
[sbt212] [info]   sbt.internal.inc.InvalidComponent: The compiler bridge sources org.scala-sbt:compiler-bridge_2.10:1.3.0-bin-20190418T141950:compile could not be retrieved.
```

*head scratch*

**Comment by eed3si9n** [comment]
Is the bridge published?

**Comment by dwijnand** [comment]
I guess not?  The head scratching comes from why it wants the 2.13 bridge in the 2.10 test case.  I'll iterate locally and fix the tests (I only iterated on tests compiling).

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1554/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/4617/head-develop-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1554/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@044ced544fa2422fa4761efccff17f9b207db756
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@a5c269c729100140981e1c9ea49568a256fa92f2
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@dda58a019335c1d8480a516bdd1e97ede0b7bdcb
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=develop&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by dwijnand** [comment]
```
[sbt212] [error] /home/jenkinssbt/workspace/sbt-validator/target-0.9.16/project-builds/sbt212-7282eaf24b43dd9e598d0d45a868acfe5552410f/internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala:90:26: not found: value ScrollUp
[sbt212] [error]     if (showProgress) s"$ScrollUp$DeleteLine" + prompt
[sbt212] [error]                          ^
[sbt212] [error] /home/jenkinssbt/workspace/sbt-validator/target-0.9.16/project-builds/sbt212-7282eaf24b43dd9e598d0d45a868acfe5552410f/internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala:90:26: not found: value ScrollUp
[sbt212] [error]     if (showProgress) s"$ScrollUp$DeleteLine" + prompt
[sbt212] [error]                          ^
[sbt212] [info] No documentation generated with unsuccessful compiler run
[sbt212] [error] one error found
[sbt212] [error] one error found
[sbt212] [error] (completeProj / Compile / compileIncremental) Compilation failed
[sbt212] [error] (completeProj / Compile / doc) Scaladoc generation failed
```

weird

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1555/console">is in progress...</a>

**Comment by dwijnand** [comment]
Rebased and re-ran sbt-validator.

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/4617/head-develop-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1555/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@15e9356e27958f669c04e8140031652c1e5d8738
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@a5c269c729100140981e1c9ea49568a256fa92f2
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@dda58a019335c1d8480a516bdd1e97ede0b7bdcb
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=develop&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by dwijnand** [comment]
🙄

```
[sbt213-io] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt213-io] [warn] 	::          UNRESOLVED DEPENDENCIES         ::
[sbt213-io] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[sbt213-io] [warn] 	:: org.scalatest#scalatest_2.13.0-M5;3.0.8-RC2: not found
[sbt213-io] [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
```

The rest of it was successful:

```
[info] Project sbt210-io-------------------: SUCCESS (project rebuilt ok)
[info] Project sbt210-util-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt210-zinc-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt211-io-------------------: SUCCESS (project rebuilt ok)
[info] Project sbt211-librarymanagement----: SUCCESS (project rebuilt ok)
[info] Project sbt211-util-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt211-zinc-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt212----------------------: SUCCESS (project rebuilt ok)
[info] Project sbt212-io-------------------: SUCCESS (project rebuilt ok)
[info] Project sbt212-librarymanagement----: SUCCESS (project rebuilt ok)
[info] Project sbt212-util-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt212-util-interface-------: SUCCESS (project rebuilt ok)
[info] Project sbt212-zinc-----------------: SUCCESS (project rebuilt ok)
[info] Project sbt213-io-------------------: FAILED (RuntimeException: )
[info] Project sbt213-zinc-----------------: SUCCESS (project rebuilt ok)
[info] Project scala-210-------------------: SUCCESS (unchanged, not rebuilt)
[info] Project scala-211-------------------: SUCCESS (unchanged, not rebuilt)
[info] Project scala-212-------------------: SUCCESS (unchanged, not rebuilt)
[info] Project scala-213-------------------: SUCCESS (unchanged, not rebuilt)
[info] >>> The dbuild result is----------------: FAILED (failed: sbt213-io)
```

Ok, I'll recover the commit history now.

**Comment by eed3si9n** [comment]
```
[info] Project sbt213-io-------------------: FAILED (RuntimeException: )
```

This is likely due to sbt-validator using Scala 2.13.0-M5 instead of RC1.

**Comment by eed3si9n** [comment]
https://github.com/sbt/sbt-validator/commit/7071546c9344a5ff29ea34520a51998e04f89635

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1556/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: pull/4617/head-develop-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1556/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4617">pull/4617/head      </A> | sbt/sbt@15e9356e27958f669c04e8140031652c1e5d8738
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@a5c269c729100140981e1c9ea49568a256fa92f2
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@dda58a019335c1d8480a516bdd1e97ede0b7bdcb
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4617/head&zinc=develop&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by dwijnand** [comment]
Ready for review, @eed3si9n

**Review by eed3si9n — DISMISSED** [review]

**Comment by eed3si9n** [comment]
Zinc 1.3.0-M4 is out. See https://github.com/dwijnand/sbt/pull/2

File: zinc-lm-integration/src/main/scala/sbt/internal/inc/ZincComponentCompiler.scala [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    def compilerBridgeId(scalaVersion: String) = {
+      scalaVersion match {
+        case sc if (sc startsWith "2.10.")       => "compiler-bridge_2.10"
+        case sc if (sc startsWith "2.11.")       => "compiler-bridge_2.11"
+        case sc if (sc startsWith "2.12.")       => "compiler-bridge_2.12"
+        case "2.13.0-M1"                         => "compiler-bridge_2.12"
+        case sc if (sc startsWith "2.13.0-pre-") => "compiler-bridge_2.13.0-M5"
+        case sc if (sc startsWith "2.13.0-M")    => "compiler-bridge_2.13.0-M5"
+        case sc if (sc startsWith "2.13.0-RC")   => "compiler-bridge_2.13.0-M5"
+        case _                                   => "compiler-bridge_2.13"
```
Comment (eed3si9n) [inline zinc-lm-integration/src/main/scala/sbt/internal/inc/ZincComponentCompiler.scala]: should be

```
        case "2.13.0-M1"                   => "compiler-bridge_2.12"
        case _                             => "compiler-bridge_2.13"
```

https://github.com/sbt/zinc/pull/651

**Review by eed3si9n — CHANGES_REQUESTED** [review]

**Review by eed3si9n — APPROVED** [review]

**Comment by dwijnand** [comment]
Thank you, Eugene, for helping me land this.

**Pull request merged**